Fix remote_tmp when become with non admin user

jborean93 opened the pull request:
##### SUMMARY
Further changes based on review of https://github.com/ansible/ansible/pull/41568

This should fix the issue of a module trying to access the Ansible controller created remote tmp dir if become is used for a non admin user.

Fixes https://github.com/ansible/ansible/issues/41340
Fixes https://github.com/ansible/ansible/issues/42117

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
remote_tmp
basic.py

##### ANSIBLE VERSION
```
2.6+
```

bcoca committed:
Fix tmpdir on non root become

 - also avoid exception if tmpdir and remote_tmp are None
 - give 'None' on deescalation so tempfile will fallback to it's default behaviour
   and use system dirs
 - fix issue with bad tempdir (not existing/not createable/not writeable)
   i.e nobody and ~/.ansible/tmp
 - added tests for blockfile case

bcoca committed:
Revert "Temporarily revert c119d54"

This reverts commit 5c614a59a66fc75b6e258053d3d17d151141e7f9.

jborean93 committed:
changes based on PR feedback and changelog fragment

abadger commented on lib/ansible/plugins/action/__init__.py:
```diff
@@ -231,29 +231,22 @@ def _make_tmp_path(self, remote_user=None):
-            admin_users = self._connection._shell.get_option('admin_users') + [remote_user]
-        except AnsibleError:
-            admin_users = ['root', remote_user]  # plugin does not support admin_users
+        deescalated_user = self._deescalated_user()
```
Note this covers more than deescalation.

Also, in general functions should be verbs and variables should be nouns.

Maybe something like this?
```
become_unprivileged = self._is_become_unprivileged()
```

abadger commented on lib/ansible/plugins/action/__init__.py:
```diff
@@ -1033,3 +1024,33 @@ def _find_needle(self, dirname, needle):
+            remote_user = self._connection.get_option('remote_user')
+        except AnsibleError:
+            remote_user = self._play_context.remote_user
+        return bool(self._play_context.become_user not in admin_users + [remote_user])
```
In general, (except for __init__) our structure is for callers to be towards the bottom of files and callees to be towards the top.  So these should be somewhere above _make_tmp_path() in the class.

jborean93 replied:
No worries, is this just an Ansible thing or common in Python?

abadger replied:
I've seen it in Python more frequently than putting callers at the top and callees below.  I know I got this style because  I came in via Pascal, C, and C++ which all use it (At least, at the time when I learned C++).  I'm not sure if that has any bearing on why other Python programmers/projects use it but it does seem to be the most popular option.

abadger commented on lib/ansible/plugins/action/__init__.py:
```diff
@@ -1033,3 +1024,33 @@ def _find_needle(self, dirname, needle):
+        Returns a list of admin users that are configured for the current shell
+        plugin
+        '''
+        # fallback for old custom plugins w/o get_option
```
Move the comment into the except block (since it only apples to admin_users = ['root'], right?)

jborean93 replied:
it applies when a shell plugin (3rd party) doesn't have an `admin_users` option but you are right this should be in the except block.

abadger commented on lib/ansible/plugins/action/__init__.py:
```diff
@@ -1033,3 +1024,33 @@ def _find_needle(self, dirname, needle):
+        The user is not the same as the connection user and is not part of the
+        shell configured admin users
+        '''
+        # if we don't use become then we know the user is deescalated
```
I think the comment here should be the opposite:
```
# If we don't use become then we know we aren't switching to a different unprivileged user
```

jborean93 replied:
yep, will fix up.

abadger commented on lib/ansible/plugins/action/__init__.py:
```diff
@@ -1033,3 +1024,33 @@ def _find_needle(self, dirname, needle):
+            return False
+
+        # if we use become and the user is not an admin (or same user) then
+        # we need to deescalate
```
Remember to change deescalate in this comment as well.

jborean93 committed:
changes based on the review

jborean93 committed:
Fix tmpdir when makedirs failed so we just use the system tmp

abadger committed:
Let missing remote_tmp fail

If remote_tmp is missing then there's something more basic wrong in the
communication from the controller to the module-side.  It's better to
be alerted in this case than to silently ignore it.

jborean and I have independently checked what happens if the user sets
ansible_remote_tmp to empty string and !!null and both cases work fine.
(null is turned into a default value controller-side.  empty string
triggers the warning because it is probably not a directory that the
become user is able to use).

jborean93 commented:
Backport PR to stable-2.6 https://github.com/ansible/ansible/pull/42483

mcallaghan-sandvine commented:
How does I get this fix into the 2.6.X stable branch?
https://releases.ansible.com/ansible/rpm/release/fedora-27-x86_64/
